AMDGPU/GlobalISel: add RegBankLegalize rules for select

petar-avramovic opened the pull request:
Uniform condition S1 is AnyExtended to S32 and high bits are cleaned using
AND with 1. Divergent S1 uses VCC.
Using B32/B64 rules to cover scalars vector and pointer types.
Divergent B64 is split to S32.

petar-avramovic commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132384?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#132385** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132385?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#132384** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132384?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132384?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#132383** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132383?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#132382** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132382?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#132381** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132381?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

ritter-x2a commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -492,7 +511,8 @@ LLT RegBankLegalizeHelper::getBTyFromID(RegBankLLTMappingApplyID ID, LLT Ty) {
         Ty == LLT::fixed_vector(4, 16) || Ty == LLT::pointer(0, 64) ||
-        Ty == LLT::pointer(1, 64) || Ty == LLT::pointer(4, 64))
+        Ty == LLT::pointer(1, 64) || Ty == LLT::pointer(4, 64) ||
+        Ty == LLT::pointer(999, 64))
```
I don't think we should have special case handling for address space 999. The tests using that AS surely just use it to represent "any other address space"?

petar-avramovic replied:
Don't know much about the topic, would all pointer types with addr space greater then 6 all be accepted and be 64 bits?

ritter-x2a replied:
7 and 8 are bigger than that (see [here](https://llvm.org/docs/AMDGPUUsage.html#amdgpu-address-spaces-table), but I don't know if we support them in Global ISel). We treat everything above `AMDGPUAS::MAX_AMDGPU_ADDRESS` as an alias to the 64-bit flat AS [here](https://github.com/llvm/llvm-project/blob/54ff03532148bfc48ace062e453eaad8ec74fdda/llvm/include/llvm/Support/AMDGPUAddrSpace.h#L94).

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -485,7 +504,8 @@ LLT RegBankLegalizeHelper::getBTyFromID(RegBankLLTMappingApplyID ID, LLT Ty) {
         Ty == LLT::fixed_vector(4, 16) || Ty == LLT::pointer(0, 64) ||
-        Ty == LLT::pointer(1, 64) || Ty == LLT::pointer(4, 64))
+        Ty == LLT::pointer(1, 64) || Ty == LLT::pointer(4, 64) ||
+        (Ty.isPointer() && Ty.getAddressSpace() > AMDGPUAS::MAX_AMDGPU_ADDRESS))
```
What case is this trying to handle? Is it tested?

arsenm replied:
We should try to recognize any unhandled address spaces as global aliases, there's a p999 test

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -198,7 +198,8 @@ UniformityLLTOpPredicateID LLTToBId(LLT Ty) {
       Ty == LLT::fixed_vector(4, 16) || Ty == LLT::pointer(1, 64) ||
-      Ty == LLT::pointer(4, 64))
+      Ty == LLT::pointer(4, 64) ||
+      (Ty.isPointer() && Ty.getAddressSpace() > AMDGPUAS::MAX_AMDGPU_ADDRESS))
```
same question here

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -286,6 +287,22 @@ void RegBankLegalizeHelper::lowerSplitTo32(MachineInstr &MI) {
 
+void RegBankLegalizeHelper::lowerSplitTo32Sel(MachineInstr &MI) {
+  Register Dst = MI.getOperand(0).getReg();
+  LLT Ty = MRI.getType(Dst) == V4S16 ? V2S16 : S32;
```
```suggestion
  LLT Ty = (MRI.getType(Dst) == V4S16 ? V2S16 : S32);
```
nit for clarity

+ if you expect the type to be something else when it isn't `V4S16`, add an assert?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -286,6 +287,25 @@ void RegBankLegalizeHelper::lowerSplitTo32(MachineInstr &MI) {
+  LLT DstTy = MRI.getType(Dst);
+  assert(DstTy == V4S16 || DstTy == V2S32 || DstTy == S64 ||
+         (DstTy.isPointer() && DstTy.getSizeInBits() == 64));
+  LLT Ty = (DstTy == V4S16 ? V2S16 : S32);
```
```suggestion
  LLT Ty = DstTy == V4S16 ? V2S16 : S32;
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -286,6 +287,25 @@ void RegBankLegalizeHelper::lowerSplitTo32(MachineInstr &MI) {
   MI.eraseFromParent();
 }
 
+void RegBankLegalizeHelper::lowerSplitTo32Sel(MachineInstr &MI) {
```
Spell out Select

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -372,6 +392,8 @@ void RegBankLegalizeHelper::lower(MachineInstr &MI,
     return lowerS_BFE(MI);
   case SplitTo32:
     return lowerSplitTo32(MI);
+  case SplitTo32Sel:
```
Spell out select

petar-avramovic committed:
AMDGPU/GlobalISel: add RegBankLegalize rules for select

Uniform condition S1 is AnyExtended to S32 and high bits are
cleaned using AND with 1. Divergent S1 uses VCC.
Using B32/B64 rules to cover scalars vector and pointer types.
Divergent B64 is split to S32.

petar-avramovic commented:
### Merge activity

* **May 26, 10:03 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/132384).
